Runtime/Core/ConfinerOven.cs

```diff
@@ -0,0 +1,300 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Assertions;
+
+namespace Cinemachine
+{
+    /// <summary>
+    /// Responsible for baking confiners via BakeConfiner function.
+    /// </summary>
+    internal class ConfinerOven
+    {
+        public class ConfinerState
+        {
+            public List<ShrinkablePolygon> m_Polygons;
+            public float m_WindowSize;
+            public float m_State;
+        }
+        
+        private List<List<ShrinkablePolygon>> m_shrinkablePolygons;
+        
+        /// <summary>
+        /// Creates shrinkable polygons from input parameters.
+        /// The algorithm is divide and conquer. It iteratively shrinks down the input polygon towards its shrink
+        /// directions. If the polygon intersects with itself, then we divide the polygon into two polygons at the
+        /// intersection point, and continue the algorithm on these two polygons separately. We need to keep track of
+        /// the connectivity information between sub-polygons.
+        /// </summary>
+        public void BakeConfiner(in List<List<Vector2>> inputPath, in float sensorRatio, in float shrinkAmount, 
+            float maxOrthosize, in bool shrinkToPoint)
+        {
+            float polygonHalfHeight = HeightOfAspectBasedBoundingBoxAroundPolygons(inputPath, sensorRatio) / 2f;
+            if (maxOrthosize == 0 || maxOrthosize > polygonHalfHeight) // exact comparison to 0 is intentional!
+            {
+                // ensuring that we don't compute further than what is the theoretical max
+                maxOrthosize = polygonHalfHeight; 
+            }
+            
+            m_shrinkablePolygons = CreateShrinkablePolygons(inputPath, sensorRatio);
+            var polyIndex = 0;
+            var shrinking = true;
+            while (shrinking)
+            {
+                var nextPolygonIteration = new List<ShrinkablePolygon>();
+                for (int g = 0; g < m_shrinkablePolygons[polyIndex].Count; ++g)
+                {
+                    m_shrinkablePolygons[polyIndex][g].ComputeAspectBasedShrinkDirections();
+                    ShrinkablePolygon shrinkablePolygon = m_shrinkablePolygons[polyIndex][g].DeepCopy();
+                    if (shrinkablePolygon.Shrink(shrinkAmount, shrinkToPoint))
+                    {
+                        if (shrinkablePolygon.m_WindowDiagonal > shrinkAmount * 100f)
```

Comment:
what is this 100 constant?  Is it needed because shrinkAmount is so small?

Reply:
It represent 100 steps, but it was arbitrary.